app/javascript/app/service-provider-form.js

```diff
@@ -92,13 +92,83 @@ function protocolOptionSetup() {
   });
 }
 
+function certificateUploadSetup() {
+  // Selectors
+  const fileContainer = document.getElementById('certificate-container');
+  const pemInputMessage = document.querySelector('.js-pem-input-error-message');
+  const pemInput = document.querySelector('.js-pem-input');
+  const pemFilename = document.querySelector('.js-pem-file-name');
+
+  // Functions
+  const setPemError = (message) => {
+    pemInputMessage.textContent = message;
+  };
+
+  const handleUploadedCert = () => {
+    const file = pemInput.files[0];
+
+    pemFilename.textContent = file ? file.name : null;
+
+    if (file && file.text) {
+      file.text().then((content) => {
+        if (content.includes('PRIVATE')) {
+          setPemError('This is a private key, upload the public key instead');
+        } else if (!content.includes('-----BEGIN CERTIFICATE-----')) {
+          setPemError('This file does not appear to be PEM encoded');
+        } else {
+          setPemError(null);
+        }
+      });
+    } else {
+      setPemError(null);
+    }
+  };
+
+  // Event triggers
+  fileContainer.addEventListener('change', handleUploadedCert);
+}
+
+function logoUploadSetup() {
+  // Selectors
+  const logoInput = document.querySelector('.input-file');
+  const logoPreview = document.querySelector('.input-preview');
+
+  // Funcitons
+  const handleUploadedLogo = () => {
+    const logoFile = logoInput.files[0];
+
+    logoPreview.textContent = (logoFile && logoFile.name) || '';
+  };
+
+  // Event triggers
+  logoInput.addEventListener('change', handleUploadedLogo);
+}
+
+function redirectURISetup() {
+  // Selectors
+  const redirectURI = document.getElementById('add-redirect-uri-field');
+
+  // Functions
+  const handleRedirectURIClick = () => {
+    const lastInput = document.querySelector('.service_provider_redirect_uris input:last-child');
+    const newInput = lastInput.cloneNode(true);
+    newInput.value = '';
```

Comment:
Are we meant to be doing something with this new input? Appending it to something?

Reply:
So, I was confused about what `cloneNode` did. When I appended I was seeing 2 inputs show up. I realize now that was because the old code was in there adding an input. I'll put the code to append it back and remove the old code 😳